grape/fault_diagnosis.py

```diff
@@ -120,6 +119,143 @@ def check_before(self):
+
+        for switch, opened in zip(initial_condition.keys(), individual):
+            if not opened: # we remove edges if the switch is set to 'False'
+                for succ in list(T.successors(switch)):
+                    T.remove_edge(switch, succ)
+
+        for node in perturbed_nodes:
+            if node in T.nodes():
+
+                _, broken_nodes = self.rm_nodes(node, T)
+                broken_nodes = list(set(broken_nodes))
+
+                for n in broken_nodes: T.remove_node(n)
+
+        return (n_actions - sum(T.service.values())- len(T),)
+
+    def optimizer(self, perturbed_nodes, initial_condition, params):
+        """
+
+        Genetic algorithm to optimize switches conditions, using DEAP.
+
+        :param list perturbed_nodes: nodes(s) involved in the
+            perturbing event.
+        :param dict initial_condition: initial status (boolean) for the graph
+            switches.
+        :param dict params: values for the optimizer evolutionary algorithm.
+            Dict of: {str: int, str: int, str: float}.
+
+            'npop' -- number of individuals for each population (default to 300)
+            'ngen' -- total number of generations (default to 100)
+            'indpb' -- independent probability for attributes to be changed
+                (default to 0.6) 
+        """
+
+        logging.getLogger().setLevel(logging.INFO)
+
+        creator.create("FitnessMin", base.Fitness, weights=(-1.0,))
+        creator.create("Individual", list, fitness=creator.FitnessMin)
+    
+        toolbox = base.Toolbox()
+        # Attribute generator
+        toolbox.register("attribute_bool", random.choice, [True, False])
+        # Structure initializers
+        toolbox.register("individual", tools.initRepeat, creator.Individual,
+            toolbox.attribute_bool, len(self.G.switches))
+        toolbox.register("population", tools.initRepeat, list,
+            toolbox.individual)
+    
+        toolbox.register("evaluate", self.fitness_evaluation)
+        toolbox.register("mate", tools.cxUniform, indpb=params['indpb'])
+        toolbox.register("mutate", tools.mutShuffleIndexes,
+            indpb=params['indpb'])
+        toolbox.register("select", tools.selBest)
+
+        pop = toolbox.population(n=params['npop'])
+        # Evaluate the entire population
+        fitnesses = [toolbox.evaluate(ind, perturbed_nodes,
+            initial_condition) for ind in pop]
+
+        for ind, fit in zip(pop, fitnesses):
+            ind.fitness.values = fit
+        fits = [ind.fitness.values[0] for ind in pop]
+
+        # Variable keeping track of the number of generations
+        g = 0
+    
+        fitnesses = fits
+        result = []
+        # Begin the evolution
+
+        while g < params['ngen']:
+            # A new generation
+            g = g + 1
+
+            # Select the next generation individuals
+            offspring = toolbox.select(pop, 5)
```

Comment:
I would move the hardcoded threshold to the beginning of the method (maybe in the future also to `param` to give the user the possibility to change such value)